### PR TITLE
Orphaned item cleanup

### PR DESCRIPTION
**Pull request by Ebag333**
See #954 for some history.

Here's an example output from cleaning up a very large and very old database.

```
2017-01-27 13:36:46,260 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for orphaned drones items.
2017-01-27 13:36:46,306 eos.db.saveddata.databaseRepair ERROR    Database corruption found. Cleaning up 21 records.
2017-01-27 13:36:46,315 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for orphaned cargo items.
2017-01-27 13:36:46,357 eos.db.saveddata.databaseRepair ERROR    Database corruption found. Cleaning up 3 records.
2017-01-27 13:36:46,358 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for orphaned fighters items.
2017-01-27 13:36:46,362 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for orphaned modules items.
2017-01-27 13:36:46,517 eos.db.saveddata.databaseRepair ERROR    Database corruption found. Cleaning up 3251 records.
2017-01-27 13:36:46,519 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for null damagePatterns values. (em)
2017-01-27 13:36:46,548 eos.db.saveddata.databaseRepair ERROR    Database corruption found. Cleaning up 1 records.
2017-01-27 13:36:46,549 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for null damagePatterns values. (thermal)
2017-01-27 13:36:46,576 eos.db.saveddata.databaseRepair ERROR    Database corruption found. Cleaning up 1 records.
2017-01-27 13:36:46,578 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for null damagePatterns values. (kinetic)
2017-01-27 13:36:46,611 eos.db.saveddata.databaseRepair ERROR    Database corruption found. Cleaning up 2 records.
2017-01-27 13:36:46,612 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for null damagePatterns values. (explosive)
2017-01-27 13:36:46,653 eos.db.saveddata.databaseRepair ERROR    Database corruption found. Cleaning up 2 records.
2017-01-27 13:36:46,654 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for null targetResists values. (em)
2017-01-27 13:36:46,660 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for null targetResists values. (thermal)
2017-01-27 13:36:46,664 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for null targetResists values. (kinetic)
2017-01-27 13:36:46,668 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for null targetResists values. (explosive)
```

Primarily we're cleaning up cargo, drones, fighters, and modules that no longer have either a `itemID` or `fitID` (so the item doesn't exist, or the fit doesn't exist).

We're also catching any damage or target profiles that have a null value, and setting it to 0.

Finally, we're catching a scenario where multiple `Selected Ammo` profiles can exist.

### CODE REVIEW

**Comment by Ebag333** [comment]
@blitzmann this is based on #953 so you'll probably want to merge that one first.

This should be ready to merge.

**Comment by blitzmann** [comment]
If the item doesn't exist in the eve database, we already have a mechanism in pace for removing it from a fit. Likewise, if a module is orphaned from a fit, there shouldn't be any negative effects other than keeping that data around for no reason.

I'll take a look at the other stuff when I can, but the two that I mentioned shouldn't need to be handled, at least not at every start up.

**Comment by Ebag333** [comment]
Shouldn't, but there's an atrocious number of errors in the log that indicates that there is.  Most likely them mechanism for cleaning up items isn't working all the time, and since these are genuinely orphaned items there's no harm in removing them.  If nothing else it'll help keep the DB from growing a bit.  And since we're checking for both missing fit ID and item ID at the same time, if we're doing one might as well do the other (essentially the same amount of time regardless).

```2017-01-27 13:36:46,362 eos.db.saveddata.databaseRepair DEBUG    Running database cleanup for orphaned modules items.
2017-01-27 13:36:46,517 eos.db.saveddata.databaseRepair ERROR    Database corruption found. Cleaning up 3251 records.
```

Even running in full debugging mode and removing a atrocious number of orphaned items this takes essentially no time at all.

With a clean DB (again in full potato mode) and running all the old and new checks:
```
2017-01-27 13:57:14,375 root                     DEBUG    Starting database validation.
2017-01-27 13:57:14,490 root                     DEBUG    Completed database validation.
```
we're taking just over 100 ms to run every check.  That's not a significant amount of time at all, there's probably plenty of other things we could do to get that time back and then some (like threading fit calculation as per #848).

**Comment by blitzmann** [comment]
Oh it's just checking for a null, blank, or 0 item id. I thought it was doing lookups to the eve database as well.

Mind sending the database over to me so I can take a peek? As always, it's like to fix the root cause of possible - there shouldn't be any reason to have everything blank or 0 vales for item ID  (item ID can be null if it's an empty slot iirc,, in which case we shouldn't touch them)

**Comment by Ebag333** [comment]
>(item ID can be null if it's an empty slot iirc,, in which case we shouldn't touch them)

And of course modules have to be a special snowflake. :)

Shouldn't _hurt_ it by nuking it, but I'll tweak the logic to avoid touching modules in that way.

**Comment by Ebag333** [comment]
>I'll tweak the logic to avoid touching modules in that way.

FYI this is done @blitzmann

File: eos/db/saveddata/databaseRepair.py [hunk 1]
```diff
@@ -160,3 +160,75 @@ def NullTargetResistNames(saveddata_engine):
...
+                query = "SELECT COUNT(*) AS num FROM " + profileType + " WHERE " + damageType + "Amount IS NULL OR " + damageType + "Amount = ''"
+                results = DatabaseCleanup.ExecuteSQLQuery(saveddata_engine, query)
+
+                if results is None:
+                    return
+
+                row = results.first()
+
+                if row and row['num']:
+                    query = "UPDATE '" + profileType + "' SET '" + damageType + "Amount' = '0' WHERE " + damageType + "Amount IS NULL OR emAmount = ''"
```
Comment (blitzmann) [inline eos/db/saveddata/databaseRepair.py]: looks like `emAmount` was left here, but it's easy enough for me to change after merging.`

Comment (Ebag333) [inline eos/db/saveddata/databaseRepair.py]: oops yeah just drop t he `em` part of it.

**Pull request merged**